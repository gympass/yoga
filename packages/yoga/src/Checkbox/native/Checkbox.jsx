import React, { useState } from 'react';
import { bool, func, string, oneOf, shape } from 'prop-types';
import styled, { withTheme } from 'styled-components';
import { TouchableWithoutFeedback, View } from 'react-native';
import { hexToRgb } from '@gympass/yoga-common';
import { Check } from '@gympass/yoga-icons';

const CheckboxWrapper = styled.View`
  flex-direction: row;
  align-items: center;
`;

const Label = styled.Text(
  ({
    theme: {
      yoga: {
        components: { checkbox },
      },
    },
  }) => `
  padding-left: ${checkbox.label.padding.left}px;

  font-size: ${checkbox.label.font.size}px;

  color: ${checkbox.label.font.color};
`,
);

const HelperWrapper = styled.View(
  ({
    theme: {
      yoga: {
        components: { checkbox },
      },
    },
  }) => `
  margin-top: ${checkbox.helper.margin.top}px;
`,
);

const Helper = styled.Text(
  ({
    error,
    theme: {
      yoga: {
        colors: { negative },
        components: { checkbox },
      },
    },
  }) => `
  font-size: ${checkbox.helper.font.size}px;

  color: ${error ? negative[1] : checkbox.helper.font.color};
`,
);

const CheckBackground = styled.View(
  ({
    checked,
    disabled,
    error,
    variant,
    theme: {
      yoga: {
        colors: { [variant]: color = [], negative },
        components: { checkbox },
      },
    },
  }) => `
    position: absolute;

    width: ${checkbox.size}px;
    height: ${checkbox.size}px;

    border-radius: ${checkbox.border.radius}px;
    border-width: ${checkbox.border.width}px;
    border-style: solid;

    ${checked ? `background-color: ${color[3]};` : ''}

    border-color: ${disabled ? checkbox.disabled.border.color : color[3]};

    ${
      disabled && checked
        ? `background-color: ${checkbox.disabled.backgroundColor};`
        : ''
    }

    ${error ? `border-color: ${negative[1]};` : ''}

    ${error && checked ? `background-color: ${negative[1]};` : ''}
  `,
);

const Shadow = styled.View(
  ({
    variant,
    theme: {
      yoga: {
        colors: { [variant]: color = [] },
        components: { checkbox },
      },
    },
  }) => {
    const size = checkbox.size * 1.67;

    return `
      width: ${size}px;
      height: ${size}px;

      background-color: ${hexToRgb(color[1], 0.75)};          
      border-radius: ${checkbox.hover.border.radius}px;
    `;
  },
);

const CheckArea = styled.View(
  ({
    theme: {
      yoga: {
        components: { checkbox },
      },
    },
  }) => `
  justify-content: center;
  align-items: center;

  width: ${checkbox.size}px;
  height: ${checkbox.size}px;
`,
);

/** The checkbox component is used when the user needs to select one or more
 * items on a task. This component can also allow the user to turn an option on
 * or off.  */
const Checkbox = ({
  label,
  helper,
  disabled,
  checked,
  error,
  style,
  variant,
  onPressIn,
  onPressOut,
  theme: {
    yoga: {
      components: { checkbox },
    },
  },
  ...rest
}) => {
  const [pressed, setPressed] = useState(false);
  return (
    <View>
      <TouchableWithoutFeedback
        {...rest}
        disabled={disabled}
        onPressIn={e => {
          setPressed(true);
          onPressIn(e);
        }}
        onPressOut={e => {
          setPressed(false);
          onPressOut(e);
        }}
      >
        <CheckboxWrapper style={style}>
          <CheckArea>
            {pressed && !disabled && (
              <Shadow pressed={pressed} variant={variant} />
            )}
            <CheckBackground
              {...{
                disabled,
                checked,
                error,
                pressed,
                variant,
              }}
            />
            {checked && (
              <Check
                fill={checkbox.checked.icon.color}
                style={{ position: 'absolute' }}
              />
            )}
          </CheckArea>

          {label && <Label>{label}</Label>}
        </CheckboxWrapper>
      </TouchableWithoutFeedback>
      {(helper || error) && (
        <HelperWrapper>
          <Helper error={error}>{error || helper}</Helper>
        </HelperWrapper>
      )}
    </View>
  );
};

Checkbox.propTypes = {
  label: string,
  /** set a short helper text under checkbox */
  helper: string,
  checked: bool,
  disabled: bool,
  error: string,
  style: shape({}),
  /** style the card following the theme (primary, secondary, tertiary) */
  variant: oneOf(['primary', 'secondary', 'tertiary']),
  onPressIn: func,
  onPressOut: func,
};

Checkbox.defaultProps = {
  label: undefined,
  helper: undefined,
  checked: false,
  disabled: false,
  error: undefined,
  style: {},
  variant: 'primary',
  onPressIn: () => {},
  onPressOut: () => {},
};

Checkbox.displayName = 'Checkbox';

export default withTheme(Checkbox);
