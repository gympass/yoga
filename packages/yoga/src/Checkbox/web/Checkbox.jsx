import React, { useEffect, useRef, useId } from 'react';
import { bool, string, shape, oneOfType, node } from 'prop-types';
import styled, { withTheme } from 'styled-components';
import { hexToRgb } from '@gympass/yoga-common';
import { Check, Rectangle } from '@gympass/yoga-icons';

import { HiddenInput } from '../../shared';

const CheckboxWrapper = styled.div`
  display: inline-block;

  * {
    box-sizing: border-box;
  }

  ${({ disabled }) => (disabled ? `cursor: not-allowed` : '')}
`;

const CheckMark = styled.div`
  position: relative;

  border-style: solid;

  flex-shrink: 0;

  ${({
    checked,
    disabled,
    inverted,
    error,
    indeterminate,
    theme: {
      yoga: {
        colors: { primary, feedback, elements, white },
        components: { checkbox },
      },
    },
  }) => {
    let borderColor = elements.selectionAndIcons;
    let bgColor = 'transparent';
    let checkColor = checkbox.checked.icon.color;

    if (error) {
      [, borderColor] = feedback.attention;

      if (checked || indeterminate) {
        [, bgColor] = feedback.attention;
      }
    } else if (disabled) {
      borderColor = checkbox.disabled.border.color;

      if (checked || indeterminate) {
        bgColor = checkbox.disabled.backgroundColor;
        borderColor = elements.lineAndBorders;
      }
    } else if (checked || indeterminate) {
      borderColor = primary;
      bgColor = primary;
      checkColor = checkbox.checked.icon.color;

      if (inverted) {
        bgColor = white;
        borderColor = white;
        checkColor = primary;
      }
    } else if (inverted) {
      borderColor = white;
      checkColor = primary;
    }

    return `
    width: ${checkbox.size}px;
    height: ${checkbox.size}px;

    margin-right: ${checkbox.margin.right}px;

    background-color: ${bgColor};

    border-radius: ${checkbox.border.radius}px;
    border-width: ${checkbox.border.width}px;
    border-color: ${borderColor};

    svg {
      position: absolute;
      top: 50%;
      left: 50%;

      fill: ${checkColor};

      transform: translate(-50%, -50%);
    }`;
  }}
`;

const Label = styled.label`
  display: flex;
  flex-direction: row;
  align-items: center;

  cursor: pointer;

  ${({
    theme: {
      yoga: {
        components: { checkbox },
      },
    },
  }) => `
    font-size: ${checkbox.label.font.size}px;
    color: ${checkbox.label.font.color};
  `}
`;

const Shadow = styled.span`
  position: absolute;
  display: none;

  ${({
    theme: {
      yoga: {
        components: { checkbox },
      },
    },
  }) => `
    width: ${checkbox.size}px;
    height: ${checkbox.size}px;
    border-radius: ${checkbox.hover.border.radius}px;
  `}
`;

const CheckboxStyled = styled.div`
  position: relative;
  display: flex;
  align-items: center;

  ${({
    checked,
    inverted,
    disabled,
    indeterminate,
    theme: {
      yoga: {
        colors: { primary, elements, white },
        components: { checkbox },
      },
    },
  }) => {
    const shadowSize = Math.round(checkbox.size * 0.33);
    const shadowColor = () => {
      if (inverted) {
        return white;
      }

      return checked || indeterminate
        ? primary
        : elements.backgroundAndDisabled;
    };

    return `
      ${disabled ? `pointer-events: none` : ''}

      ${Label}:active,
      &:focus-within,
      &:hover  {
        ${Shadow} {
          display: block;
        }
      }

      &:hover {
        ${Shadow} {
          background-color: ${hexToRgb(shadowColor(), 0.25)};

          box-shadow: 0 0 0 ${shadowSize}px
            ${hexToRgb(shadowColor(), 0.25)};
        }
      }

      &:focus-within {
        ${Shadow} {
          background-color: ${hexToRgb(shadowColor(), 0.5)};

          box-shadow: 0 0 0 ${shadowSize}px
            ${hexToRgb(shadowColor(), 0.5)};
        }
      }

      ${Label}:active {
        ${Shadow} {
          background-color: ${hexToRgb(shadowColor(), 0.75)};

          box-shadow: 0 0 0 ${shadowSize}px
            ${hexToRgb(shadowColor(), 0.75)};
        }
      }
    `;
  }}
`;

const HelperWrapper = styled.div`
  width: 100%;

  ${({
    theme: {
      yoga: {
        components: { checkbox },
      },
    },
  }) => `
    margin-top: ${checkbox.helper.margin.top}px;
  `}
`;

const Helper = styled.span`
  ${({
    error,
    theme: {
      yoga: {
        colors: { feedback },
        components: { checkbox },
      },
    },
  }) => `
    color: ${error ? feedback.attention[1] : checkbox.helper.font.color};

    font-size: ${checkbox.helper.font.size}px;
  `}
`;

/** The checkbox component is used when the user needs to select one or more
 * items on a task. This component can also allow the user to turn an option on
 * or off.  */
const Checkbox = ({
  value,
  label,
  helper,
  disabled,
  checked,
  error,
  style,
  className,
  inverted,
  indeterminate,
  ariaLabel,
  theme: {
    yoga: {
      components: { checkbox },
    },
  },
  ...rest
}) => {
  const inputRef = useRef(null);
  const id = useId();
  const checkboxLabelId = `checkbox-label-${id}`;

  const { onChange, onClick, ...restWithoutEvents } = rest;

  useEffect(() => {
    if (inputRef.current) {
      inputRef.current.indeterminate = indeterminate;
    }
  });

  return (
    <CheckboxWrapper
      style={style}
      className={className}
      disabled={disabled}
      {...restWithoutEvents}
    >
      <CheckboxStyled
        checked={checked}
        indeterminate={indeterminate}
        inverted={inverted}
        disabled={disabled}
      >
        <Label id={checkboxLabelId}>
          <Shadow />
          <CheckMark
            {...{
              disabled,
              checked,
              error,
              inverted,
              indeterminate,
            }}
          >
            {checked && !indeterminate && (
              <Check width={checkbox.size} height={checkbox.size} />
            )}
            {indeterminate && (
              <Rectangle width={checkbox.size} height={checkbox.size} />
            )}
          </CheckMark>
          <HiddenInput
            type="checkbox"
            ref={inputRef}
            checked={checked}
            disabled={disabled}
            {...(value ? { value } : {})}
            {...restWithoutEvents}
            onChange={onChange}
            onClick={onClick}
            aria-labelledby={ariaLabel ? undefined : checkboxLabelId}
            aria-label={ariaLabel || undefined}
          />
          {label}
        </Label>
      </CheckboxStyled>
      {(helper || error) && (
        <HelperWrapper>
          <Helper error={error}>{error || helper}</Helper>
        </HelperWrapper>
      )}
    </CheckboxWrapper>
  );
};

Checkbox.propTypes = {
  label: oneOfType([string, node]),
  /** short helper text under checkbox */
  helper: string,
  value: string,
  checked: bool,
  disabled: bool,
  inverted: bool,
  error: string,
  /** if true, the component appears indeterminate */
  indeterminate: bool,
  /** set a style to the checkbox container */
  style: shape({}),
  className: string,
  ariaLabel: string,
};

Checkbox.defaultProps = {
  label: undefined,
  value: undefined,
  helper: undefined,
  checked: false,
  disabled: false,
  inverted: false,
  indeterminate: false,
  error: undefined,
  style: undefined,
  className: undefined,
  ariaLabel: undefined,
};

Checkbox.displayName = 'Checkbox';

export default withTheme(Checkbox);
