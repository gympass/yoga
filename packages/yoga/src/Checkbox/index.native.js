import { Checkbox, Switch } from './native';

const ExportCheckbox = Object.assign(Checkbox, {
  Switch,
});

export default ExportCheckbox;
