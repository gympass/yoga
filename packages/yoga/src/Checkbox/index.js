import { Checkbox, Switch } from './web';

const ExportCheckbox = Object.assign(Checkbox, {
  Switch,
});

export default ExportCheckbox;
