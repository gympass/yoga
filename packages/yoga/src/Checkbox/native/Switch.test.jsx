import React from 'react';
import { render, fireEvent } from '@testing-library/react-native';

import Checkbox from '..';
import ThemeProvider from '../../Theme';

jest.useFakeTimers();

describe('<Checkbox />', () => {
  describe('<Checkbox.Switch />', () => {
    describe('Snapshots', () => {
      it('should match snapshot with default switch', () => {
        const { toJSON } = render(
          <ThemeProvider>
            <Checkbox.Switch />
          </ThemeProvider>,
        );

        expect(toJSON()).toMatchSnapshot();
      });

      it('should match snapshot with disabled switch', () => {
        const { toJSON } = render(
          <ThemeProvider>
            <Checkbox.Switch disabled />
          </ThemeProvider>,
        );

        expect(toJSON()).toMatchSnapshot();
      });

      it('should match snapshot with checked switch', () => {
        const { toJSON } = render(
          <ThemeProvider>
            <Checkbox.Switch checked />
          </ThemeProvider>,
        );

        expect(toJSON()).toMatchSnapshot();
      });

      it('should match snapshot with disabled and checked switch', () => {
        const { toJSON } = render(
          <ThemeProvider>
            <Checkbox.Switch disabled checked />
          </ThemeProvider>,
        );

        expect(toJSON()).toMatchSnapshot();
      });

      it('should call onChange function when click on Checkbox.Switch', () => {
        const onChangeMock = jest.fn();
        const { getByRole } = render(
          <ThemeProvider>
            <Checkbox.Switch
              onChange={onChangeMock}
              accessibilityRole="button"
            />
          </ThemeProvider>,
        );

        fireEvent.press(getByRole('button'));

        expect(onChangeMock).toHaveBeenCalled();
      });
    });
  });
});
