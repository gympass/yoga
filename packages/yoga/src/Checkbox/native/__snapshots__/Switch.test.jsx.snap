// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Checkbox /> <Checkbox.Switch /> Snapshots should match snapshot with checked switch 1`] = `
<View
  accessible={true}
  focusable={false}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "flexDirection": "row",
    }
  }
>
  <View
    checked={true}
    collapsable={false}
    disabled={false}
    style={
      Object {
        "backgroundColor": "rgba(152, 152, 166, 1)",
        "borderRadius": 9999,
        "display": "flex",
        "height": 24,
        "justifyContent": "center",
        "width": 48,
      }
    }
  >
    <View
      bgColor="#FFFFFF"
      borderRadius={9999}
      checked={true}
      collapsable={false}
      disabled={false}
      elevation="0 2px 4px      rgba(152, 152, 166, 0.2)"
      height={20}
      style={
        Object {
          "backgroundColor": "#FFFFFF",
          "borderRadius": 9999,
          "height": 20,
          "shadowColor": "rgba(152, 152, 166, 0.2)",
          "shadowOffset": Object {
            "height": 2,
            "width": 0,
          },
          "shadowOpacity": 1,
          "shadowRadius": 4,
          "transform": Array [
            Object {
              "translateX": 2,
            },
          ],
          "width": 20,
        }
      }
      width={20}
    />
  </View>
</View>
`;

exports[`<Checkbox /> <Checkbox.Switch /> Snapshots should match snapshot with default switch 1`] = `
<View
  accessible={true}
  focusable={false}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "flexDirection": "row",
    }
  }
>
  <View
    checked={false}
    collapsable={false}
    disabled={false}
    style={
      Object {
        "backgroundColor": "rgba(216, 56, 94, 1)",
        "borderRadius": 9999,
        "display": "flex",
        "height": 24,
        "justifyContent": "center",
        "width": 48,
      }
    }
  >
    <View
      bgColor="#FFFFFF"
      borderRadius={9999}
      checked={false}
      collapsable={false}
      disabled={false}
      elevation="0 2px 4px      rgba(152, 152, 166, 0.2)"
      height={20}
      style={
        Object {
          "backgroundColor": "#FFFFFF",
          "borderRadius": 9999,
          "height": 20,
          "shadowColor": "rgba(152, 152, 166, 0.2)",
          "shadowOffset": Object {
            "height": 2,
            "width": 0,
          },
          "shadowOpacity": 1,
          "shadowRadius": 4,
          "transform": Array [
            Object {
              "translateX": 26,
            },
          ],
          "width": 20,
        }
      }
      width={20}
    />
  </View>
</View>
`;

exports[`<Checkbox /> <Checkbox.Switch /> Snapshots should match snapshot with disabled and checked switch 1`] = `
<View
  accessibilityState={
    Object {
      "disabled": true,
    }
  }
  accessible={true}
  focusable={false}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "flexDirection": "row",
    }
  }
>
  <View
    checked={true}
    collapsable={false}
    disabled={true}
    style={
      Object {
        "backgroundColor": "#F5F5FA",
        "borderRadius": 9999,
        "display": "flex",
        "height": 24,
        "justifyContent": "center",
        "width": 48,
      }
    }
  >
    <View
      bgColor="#FFFFFF"
      borderRadius={9999}
      checked={true}
      collapsable={false}
      disabled={true}
      elevation="0 2px 4px      rgba(152, 152, 166, 0.2)"
      height={20}
      style={
        Object {
          "backgroundColor": "#FFFFFF",
          "borderRadius": 9999,
          "height": 20,
          "shadowColor": "rgba(152, 152, 166, 0.2)",
          "shadowOffset": Object {
            "height": 2,
            "width": 0,
          },
          "shadowOpacity": 1,
          "shadowRadius": 4,
          "transform": Array [
            Object {
              "translateX": 2,
            },
          ],
          "width": 20,
        }
      }
      width={20}
    />
  </View>
</View>
`;

exports[`<Checkbox /> <Checkbox.Switch /> Snapshots should match snapshot with disabled switch 1`] = `
<View
  accessibilityState={
    Object {
      "disabled": true,
    }
  }
  accessible={true}
  focusable={false}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "flexDirection": "row",
    }
  }
>
  <View
    checked={false}
    collapsable={false}
    disabled={true}
    style={
      Object {
        "backgroundColor": "#F5F5FA",
        "borderRadius": 9999,
        "display": "flex",
        "height": 24,
        "justifyContent": "center",
        "width": 48,
      }
    }
  >
    <View
      bgColor="#FFFFFF"
      borderRadius={9999}
      checked={false}
      collapsable={false}
      disabled={true}
      elevation="0 2px 4px      rgba(152, 152, 166, 0.2)"
      height={20}
      style={
        Object {
          "backgroundColor": "#FFFFFF",
          "borderRadius": 9999,
          "height": 20,
          "shadowColor": "rgba(152, 152, 166, 0.2)",
          "shadowOffset": Object {
            "height": 2,
            "width": 0,
          },
          "shadowOpacity": 1,
          "shadowRadius": 4,
          "transform": Array [
            Object {
              "translateX": 26,
            },
          ],
          "width": 20,
        }
      }
      width={20}
    />
  </View>
</View>
`;
